src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts

```diff
@@ -556,12 +550,13 @@ export function getInputType(
             // We should be able to remove it when https://github.com/Azure/typespec-azure/issues/1733 is closed
             if (model.DiscriminatorPropertyName && m.derivedModels) {
                 for (const dm of m.derivedModels) {
-                    getInputType(
+                    const derivedModel = getInputType(
                         context,
                         getFormattedType(program, dm),
                         models,
                         enums
                     );
+                    derivedModels.push(derivedModel as InputModelType);
```

Comment:
If you update `derivedModels` here, only polymorphism with discriminator will add the derived models, other nomal inheritance will missing.
We should add dervicedModels for base model of all kinds of inheritance.

Reply:
I fixed this and add a few new test cases